Add Slack integration to Playwright E2E

eduardoumpierre opened the pull request:
Fixes #10355
Fixes #8769

#### Changes proposed in this Pull Request

This PR aims to add back the Slack integration with the E2E tests. The Slack integration was based on the [`@woocommerce/e2e-environment` package](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/e2e-environment/src/slack/reporter.js), which was the previous integration with Puppeteer.

Regarding the implementation:
- The [upload](https://api.slack.com/methods/files.upload) flow was replaced by [uploadV2](https://tools.slack.dev/node-slack-sdk/web-api/#upload-a-file) due to the regular upload flow [being deprecated](https://api.slack.com/methods/files.upload#markdown):
> files.upload is deprecated and will stop functioning on March 11, 2025. Use [files.getUploadURLExternal](https://api.slack.com/methods/files.getUploadURLExternal) and [files.completeUploadExternal](https://api.slack.com/methods/files.completeUploadExternal) to upload files instead. Newly created apps will be unable to use files.upload beginning May 8, 2024. See [Uploading files](https://api.slack.com/messaging/files#uploading_files) for more details on the process and [this changelog](https://api.slack.com/changelog/2024-04-a-better-way-to-upload-files-is-here-to-stay) for more on the deprecation.
- Only errors during the first retry will be logged to avoid spamming the Slack channel.
- A new secret was added to the GitHub actions `E2E_SLACK_CHANNEL_ID` since the current `E2E_SLACK_CHANNEL` contains a value that is not supported by the current versions of the Slack API.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* The tests should be passing
* The E2E tests should be passing
* Check the error messages in the Slack channel:
  * Error message: p1739484129382839-slack-CQ0Q6N62D
  * Screenshot message: U012K4GT0JY/F08D8R34G02-slack-file

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

eduardoumpierre committed:
Add custom Slack reporter

eduardoumpierre committed:
Temporarily adding error to E2E to validate the Slack integration

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10362` or branch name `dev/10355-playwright-slack-integration` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=dev/10355-playwright-slack-integration) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: a60c24fb25e16d99342c3df38c5e1a5327ef2409
- Build time: 2025-02-14 20:01:03 UTC

_Note: the build is updated when a new commit is pushed to this PR._

eduardoumpierre committed:
Send a Slack message only on the first test failure

eduardoumpierre committed:
Revert "Temporarily adding error to E2E to validate the Slack integration"

This reverts commit 55bd481651781a14eb8c1991425ea21d33db6e49.

eduardoumpierre committed:
Add changelog entry

alopezari reviewed (commented):
Thanks for working on this @eduardoumpierre!

As far as I know, the `e2e-environment` package, which is supported by Woo Core teams, stopped being used by them after the Playwright migration and it's only maintained as part of general repository configuration changes, but I wouldn't say it is actually supported, as it only receives the same config changes as other js packages.

In addition to the above, we just use 2 functions from this package, so keeping this dependency just for those 2 functions seems to much. If we could create our own Slack integration functions (we can copy them from this package - [sendFailedTestMessageToSlack](https://github.com/woocommerce/woocommerce/blob/d76299cb244228438fab53db7405cb44027c5c8a/packages/js/e2e-environment/src/slack/reporter.js#L82) and [sendFailedTestScreenshotToSlack](https://github.com/woocommerce/woocommerce/blob/d76299cb244228438fab53db7405cb44027c5c8a/packages/js/e2e-environment/src/slack/reporter.js#L148)) and go from there (i.e. clean them, remove unnecessary code for our case, etc), we would have one less dependency to worry about at the same time we gain more control and flexibility over this Slack integration.

What do you think?

eduardoumpierre commented:
Hey @alopezari!

That makes sense—I wasn’t aware that Woo Core teams were no longer using it, but keeping the dependency just for those two functions doesn’t seem ideal.

I’ll go ahead and follow your suggestion to copy the functions (along with the slack-web dependency) into the project. This should give us a cleaner, more integrated solution that better fits our codebase.

Thanks for checking this PR!

eduardoumpierre committed:
Add Slack web API as a dev dependency

eduardoumpierre committed:
Add custom Slack integration for E2E tests

eduardoumpierre committed:
Remove remaining Puppeteer dependencies

eduardoumpierre committed:
Temporarily adding error to E2E to validate the Slack integration

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into dev/10355-playwright-slack-integration

eduardoumpierre committed:
Update Slack integration to use the latest upload API

eduardoumpierre committed:
Update upload method to use the V2 wrapper

eduardoumpierre committed:
Use the Slack channel ID instead of its name for the Slack integration

eduardoumpierre committed:
Revert "Temporarily adding error to E2E to validate the Slack integration"

This reverts commit 3b19ef7381e23638268247b5a381a3e6d41d424d.

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into dev/10355-playwright-slack-integration

eduardoumpierre committed:
Revert "Remove remaining Puppeteer dependencies"

This reverts commit 33fab1931bbbe45c58218a3f74fe904e0ad04644.

eduardoumpierre committed:
Remove Puppeteer related dependencies

eduardoumpierre commented:
Hey @alopezari! I've updated the the code to use our custom Slack integration, removed the remaining dependencies, and updated the Slack upload flow to use the latest API `uploadV2` since the one that we were using was about to get deprecated—which also led me to add a new GitHub actions secret. The changes are described in the PR details.

The changes in the `package-lock.json` are massive, but I think that's just a lot of dependencies getting removed and the file getting reorganized.

Please, take a look at this again when you have some time. Thank you!

alopezari commented on .github/workflows/e2e-pull-request.yml:
```diff
@@ -28,6 +28,7 @@ env:
   E2E_USER_TOKEN:                ${{ secrets.E2E_USER_TOKEN }}
   WC_E2E_SCREENSHOTS:            1
   E2E_SLACK_CHANNEL:             ${{ secrets.E2E_SLACK_CHANNEL }}
```
The `E2E_SLACK_CHANNEL` was used in the `e2e-environment` package:
https://github.com/woocommerce/woocommerce/blob/8cdfbeb8a9e8f22f24a943217989f48c7091c9ce/packages/js/e2e-environment/src/slack/reporter.js#L94

Now that we're no longer using this package, can we safely remove this secret or it might be used by additional packages? It only seems to be used in E2E tests related workflows though.

eduardoumpierre replied:
I agree—since we've removed all external E2E dependencies, I’d say it's safe to remove it. I'll go ahead and update the docs and the workflows.

eduardoumpierre replied:
I've updated the code on [cb2133f](https://github.com/Automattic/woocommerce-payments/pull/10362/commits/cb2133fa90e4aa29ba781a691902bf0b7dab9ea1).

alopezari commented on .github/workflows/e2e-pw-pull-request.yml:
```diff
@@ -28,6 +28,7 @@ env:
     E2E_USER_TOKEN: ${{ secrets.E2E_USER_TOKEN }}
     WC_E2E_SCREENSHOTS: 1
     E2E_SLACK_CHANNEL: ${{ secrets.E2E_SLACK_CHANNEL }}
+    E2E_SLACK_CHANNEL_ID: ${{ secrets.E2E_SLACK_CHANNEL_ID }}
```
This secret doesn't seem to be aligned with the Slack reporting documentation for E2E tests:
https://github.com/Automattic/woocommerce-payments/blob/e792cab68320482647011ec048ac829ecac3fd89/tests/e2e/README.md#L227-L229

Are these 2 different things or we should update the documentation accordingly?

eduardoumpierre replied:
Good catch! The current Slack APIs require the Slack channel ID rather than the channel name—that's why I created the `E2E_SLACK_CHANNEL_ID` secret. I'll update the docs, since `E2E_CHANNEL_NAME` hasn't been used for a while, and also to reflect the latest state of the secrets.

eduardoumpierre replied:
I've updated the code on [cb2133f](https://github.com/Automattic/woocommerce-payments/pull/10362/commits/cb2133fa90e4aa29ba781a691902bf0b7dab9ea1). Thanks for raising it!

alopezari reviewed (approved):
Thanks for working on the new changes @eduardoumpierre!

I asked a couple of questions regarding the new slack integration.

eduardoumpierre committed:
Remove unused Slack secret and update documentation

alopezari reviewed (approved):
Thanks for adding the changes @eduardoumpierre! Looks good to me, let me approve the PR 🙌

allie500 commented on tests/e2e/utils/slack.ts:
```diff
@@ -0,0 +1,161 @@
+            The commit this build is testing is *${ commit }*. \n
+            The name of the test that failed: *${ testName }*. \n
+            See screenshot of the failed test below. ${
+				webUrl ? `*Build log* could be found here: ${ webUrl }` : ''
```
Nitpick: Can we change `could` to `can`?

eduardoumpierre replied:
Definitely! I've updated it on [d616ea1](https://github.com/Automattic/woocommerce-payments/pull/10362/commits/d616ea1209013b106c5f3d42720a09ca18def3bb). Thanks for catching it!

allie500 reviewed (approved):
Thanks for your work on this, Eduardo! The code changes and Slack message for failed test look good. I left one small nitpick for the Slack message text. Other than that, I think this is ready to ship! :shipit:

eduardoumpierre committed:
Update Slack message copy and remove unnecessary check

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into dev/10355-playwright-slack-integration
